File: app/src/main/java/com/jgm90/cloudmusic/adapters/PlaylistsAdapter.java
package com.jgm90.cloudmusic.adapters;

import android.content.Context;
import android.content.Intent;
import android.content.res.Resources;
import androidx.annotation.NonNull;
import androidx.appcompat.widget.PopupMenu;
import androidx.recyclerview.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.afollestad.materialdialogs.DialogAction;
import com.afollestad.materialdialogs.MaterialDialog;
import com.jgm90.cloudmusic.R;
import com.jgm90.cloudmusic.R2;
import com.jgm90.cloudmusic.activities.PlaylistDetailActivity;
import com.jgm90.cloudmusic.data.PlaylistData;
import com.jgm90.cloudmusic.dialogs.PlaylistDialog;
import com.jgm90.cloudmusic.interfaces.DialogCaller;
import com.jgm90.cloudmusic.interfaces.ListCaller;
import com.jgm90.cloudmusic.models.PlaylistModel;

import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;

public class PlaylistsAdapter extends RecyclerView.Adapter<PlaylistsAdapter.ViewHolder> {

    private List<PlaylistModel> model;
    private LayoutInflater inflater;
    private Context context;
    private PlaylistData dao;
    private DialogCaller dialog_listener;
    private ListCaller list_listener;

    public PlaylistsAdapter(List<PlaylistModel> model, Context context, DialogCaller listener) {
        inflater = LayoutInflater.from(context);
        this.model = model;
        this.context = context;
        this.dialog_listener = listener;
        this.dao = new PlaylistData(context);
    }

    public PlaylistsAdapter(List<PlaylistModel> model, Context context, ListCaller listener) {
        inflater = LayoutInflater.from(context);
        this.model = model;
        this.context = context;
        this.list_listener = listener;
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = inflater.inflate(R.layout.row_playlist, parent, false);
        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(ViewHolder holder, int position) {
        holder.bind(model.get(position));
    }

    @Override
    public int getItemCount() {
        return model.size();
    }

    class ViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {

        @BindView(R2.id.lbl_name)
        public TextView lbl_name;
        @BindView(R2.id.lbl_song_count)
        public TextView lbl_song_count;
        @BindView(R2.id.btn_menu)
        public ImageView btn_menu;

        public ViewHolder(View itemView) {
            super(itemView);
            ButterKnife.bind(this, itemView);
            itemView.setOnClickListener(this);
            btn_menu.setOnClickListener(this);
        }

        public void bind(PlaylistModel item) {
            lbl_name.setText(item.getName());
            Resources res = context.getResources();
            String songsFound = res.getQuantityString(R.plurals.playlist_messages, item.getSong_count(), item.getSong_count());
            lbl_song_count.setText(songsFound);
            if (list_listener != null) {
                btn_menu.setVisibility(View.INVISIBLE);
            }
        }

        @Override
        public void onClick(View v) {
            final PlaylistModel playlist = model.get(getAdapterPosition());
            if (v.getId() == btn_menu.getId()) {
                PopupMenu popup = new PopupMenu(v.getContext(), v);
                popup.inflate(R.menu.menu_playlist);
                popup.setOnMenuItemClickListener(new PopupMenu.OnMenuItemClickListener() {
                    @Override
                    public boolean onMenuItemClick(MenuItem item) {
                        int itemId = item.getItemId();
                        if (itemId == R.id.playlist_edit) {
                            new PlaylistDialog(context, dialog_listener).cargar(playlist);
                            return true;
                        } else if (itemId == R.id.playlist_delete) {
                            delete(playlist);
                            return true;
                        }
                        return true;
                    }
                });
                popup.show();
            } else {
                if (dialog_listener != null) {
                    Intent intent = new Intent(context, PlaylistDetailActivity.class);
                    intent.putExtra("PLAYLIST_ID", playlist.getPlaylist_id());
                    intent.putExtra("PLAYLIST_NAME", playlist.getName());
                    intent.putExtra("PLAYLIST_COUNT", playlist.getSong_count());
                    intent.putExtra("PLAYLIST_OFFLINE", playlist.getOffline());
                    context.startActivity(intent);
                }
                if (list_listener != null) {
                    int playlist_id = playlist.getPlaylist_id();
                    list_listener.onListItemClick(playlist_id);
                }
            }
        }

        public void delete(final PlaylistModel obj) {
            new MaterialDialog.Builder(context)
                    .title("Eliminar")
                    .content("¿Desea eliminar la lista de reproducción?")
                    .cancelable(false)
                    .positiveText("SI")
                    .onPositive(new MaterialDialog.SingleButtonCallback() {
                        @Override
                        public void onClick(@NonNull MaterialDialog dialog, @NonNull DialogAction which) {
                            dao.delete(obj);
                            dialog_listener.onPositiveCall();
                        }
                    })
                    .negativeText("NO")
                    .show();
        }
    }
}

File: app/src/main/java/com/jgm90/cloudmusic/feature/playlist/presentation/adapter/PlaylistsAdapter.kt
package com.jgm90.cloudmusic.feature.playlist.presentation.adapter

import android.content.Context
import android.content.Intent
import android.content.res.Resources
import android.view.LayoutInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.appcompat.widget.PopupMenu
import androidx.recyclerview.widget.RecyclerView
import butterknife.BindView
import butterknife.ButterKnife
import com.afollestad.materialdialogs.DialogAction
import com.afollestad.materialdialogs.MaterialDialog
import com.jgm90.cloudmusic.R
import com.jgm90.cloudmusic.R2
import com.jgm90.cloudmusic.feature.playlist.presentation.PlaylistDetailActivity
import com.jgm90.cloudmusic.feature.playlist.data.PlaylistData
import com.jgm90.cloudmusic.feature.playlist.presentation.dialogs.PlaylistDialog
import com.jgm90.cloudmusic.feature.playlist.presentation.contract.DialogCaller
import com.jgm90.cloudmusic.feature.playlist.presentation.contract.ListCaller
import com.jgm90.cloudmusic.feature.playlist.model.PlaylistModel

class PlaylistsAdapter private constructor(
    private val model: MutableList<PlaylistModel>,
    private val context: Context,
    private val dialogListener: DialogCaller?,
    private val listListener: ListCaller?,
) : RecyclerView.Adapter<PlaylistsAdapter.ViewHolder>() {
    private val inflater: LayoutInflater = LayoutInflater.from(context)
    private val dao: PlaylistData? = dialogListener?.let { PlaylistData(context) }

    constructor(model: MutableList<PlaylistModel>, context: Context, listener: DialogCaller) : this(
        model = model,
        context = context,
        dialogListener = listener,
        listListener = null,
    )

    constructor(model: MutableList<PlaylistModel>, context: Context, listener: ListCaller) : this(
        model = model,
        context = context,
        dialogListener = null,
        listListener = listener,
    )

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val view = inflater.inflate(R.layout.row_playlist, parent, false)
        return ViewHolder(view)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.bind(model[position])
    }

    override fun getItemCount(): Int = model.size

    inner class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView), View.OnClickListener {
        @JvmField
        @BindView(R2.id.lbl_name)
        var lblName: TextView? = null

        @JvmField
        @BindView(R2.id.lbl_song_count)
        var lblSongCount: TextView? = null

        @JvmField
        @BindView(R2.id.btn_menu)
        var btnMenu: ImageView? = null

        init {
            ButterKnife.bind(this, itemView)
            itemView.setOnClickListener(this)
            btnMenu?.setOnClickListener(this)
        }

        fun bind(item: PlaylistModel) {
            lblName?.text = item.name
            val res: Resources = context.resources
            val songsFound = res.getQuantityString(
                R.plurals.playlist_messages,
                item.song_count,
                item.song_count,
            )
            lblSongCount?.text = songsFound
            if (listListener != null) {
                btnMenu?.visibility = View.INVISIBLE
            }
        }

        override fun onClick(v: View) {
            val position = bindingAdapterPosition
            if (position == RecyclerView.NO_POSITION) {
                return
            }
            val playlist = model[position]
            if (v.id == btnMenu?.id) {
                val popup = PopupMenu(v.context, v)
                popup.inflate(R.menu.menu_playlist)
                popup.setOnMenuItemClickListener { item: MenuItem ->
                    when (item.itemId) {
                        R.id.playlist_edit -> {
                            PlaylistDialog(context, dialogListener).cargar(playlist)
                            true
                        }
                        R.id.playlist_delete -> {
                            delete(playlist)
                            true
                        }
                        else -> true
                    }
                }
                popup.show()
            } else {
                if (dialogListener != null) {
                    val intent = Intent(context, PlaylistDetailActivity::class.java)
                    intent.putExtra("PLAYLIST_ID", playlist.playlist_id)
                    intent.putExtra("PLAYLIST_NAME", playlist.name)
                    intent.putExtra("PLAYLIST_COUNT", playlist.song_count)
                    intent.putExtra("PLAYLIST_OFFLINE", playlist.offline)
                    context.startActivity(intent)
                }
                if (listListener != null) {
                    listListener.onListItemClick(playlist.playlist_id)
                }
            }
        }

        private fun delete(obj: PlaylistModel) {
            MaterialDialog.Builder(context)
                .title("Eliminar")
                .content("¿Desea eliminar la lista de reproducción?")
                .cancelable(false)
                .positiveText("SI")
                .onPositive(object : MaterialDialog.SingleButtonCallback {
                    override fun onClick(dialog: MaterialDialog, which: DialogAction) {
                        dao?.delete(obj)
                        dialogListener?.onPositiveCall()
                    }
                })
                .negativeText("NO")
                .show()
        }
    }
}
